### DISCUSSION TITLE
Request is throttled and retry via decorator

### DISCUSSION THREAD

**Discussion by catinthecode** [python-amazon-mws/python-amazon-mws#134] [General]
Is it possible for the make_request or MWS API call methods like Finances.list_financial_events to handle MWS throttle error?

In the old versions, I was able to wrap methods like Orders.list_orders with a decorator that retries when it catches a throttled condition, and retry certain times with increasing pause in between. But now with this new implementation of deprecating Finances.list_financial_events_by_next_token and replace it with

```
    @utils.next_token_action('ListFinancialEvents')
    def list_financial_events(self, financial_event_group_id=None, amazon_order_id=None, posted_after=None, posted_before=None, max_results=None, next_token=None)
```
Adding a retry logic via decorator does not seem to work anymore. Maybe due to how my decorator works and interfere with the decorator @utils.next_token_action('ListFinancialEvents')?

I used the decorator from this url with some minor change:

https://www.saltycrane.com/blog/2009/11/trying-out-retry-decorator-python/

Any thoughts?

**Comment by Bobspadger** [comment]
I have these working by using the Tenacity library, works a treat. 

We have discussed including this functionality in the library but we've not got there yet.

**Comment by catinthecode** [comment]
This seems a very important feature to make this library useful. I would not be able to use it without adding this retry logic. Can you add this into your feature list to be implemented?

Thanks

**Comment by Bobspadger** [comment]
@threemonks my original issue is #37 

However, I closed this as the tenacity library did this in a fantastic way and was something else for us to not have to support.

Have you checked the tenacity library out ? I can show you an example of my current working setup if you would like ?

**Comment by catinthecode** [comment]
Yes, that would be much appreciated.

**Comment by Bobspadger** [comment]
looks like I've already done it in #37  

https://github.com/python-amazon-mws/python-amazon-mws/issues/37#issuecomment-523973893

let me know if you need more help with that

**Comment by catinthecode** [comment]
Thanks. That seems would require me to create a wrapper function for each API call that could be throttled, and apply the decorator to each such wrapper function? That is doable, but would be cleaner and nicer if the throttle is built into the library code as throttling happens in almost all MWS API calls.

**Comment by Bobspadger** [comment]
Correct, you would need to wrap the API call in the tenacity decorator, but its quite easy to extract the calls from the code.

For including this in the library, as I said it has been mooted but there are other more pressing priorities first
